File: Anagrams/Anagrams.py


def anagrams(str1, str2):
    charMap1 = getCharMap(str1)
    charMap2 = getCharMap(str2)

    if len(charMap1.keys()) != len(charMap2.keys()):
        return False
    else:
        for char in charMap1.keys():
            if charMap1[char] != charMap2[char]:
                return False
        return True


def getCharMap(str):
    charMap = {}
    for char in str.replace(" ", "").lower():
        if charMap.get(char):
            charMap[char] += 1
        charMap[char] = 1
    return charMap

print(f"Hello and Olleh ---> {anagrams('Hello', 'Olleh')}")
print(f"Hello and Ollehs ---> {anagrams('Hello', 'Ollehs')}")

File: Anagrams/Anagrams.js


function anagrams (str1, str2) {
  charMap1 = buildCharMap(str1);
  charMap2 = buildCharMap(str2);

  if (Object.keys(charMap1).length !== Object.keys(charMap2).length) {
    return false;
  } else {
    for (let char in charMap1) {
      if (charMap1[char] !== charMap2[char]) return false;
    }
    return true;
  }
}


function buildCharMap (str) {
  let charMap = {}
  for (let char of str.replace(/[^\w]/g, "").toLowerCase()) {
    charMap[char] = charMap[char] + 1 || 1;
  }
  return charMap;
}


console.log(`Hello and Olleh ----> ${anagrams("Hello", "Olleh")}`);
console.log(`Hello and Ollehs ----> ${anagrams("Hello", "Ollehs")}`);
